src/Components/Web.Extensions/src/InputFile/PreFetchingSequence.cs

```diff
@@ -0,0 +1,61 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Components.Web.Extensions
+{
+    internal class PreFetchingSequence
```

Comment:
It might be worth looking at system.io.pipelines.pipe for this + the stream class, a lot of the code between the two seems to deal with the backpressure on the server, pipes already do that, so it can significantly simplify the code and save from subtle bugs.

Reply:
I initially found `System.IO.Pipelines.Pipe` a little awkward to work with and it seemed to behave about 10% slower, but it's likely I was doing something wrong - I'll give this another go.

Reply:
@javiercn I've given this a shot - let me know if you think the implementation can be improved.